File: built-in plugins dsp/mb-fm-organ.cpp
/**
 * MB FM Organ
 * Category : instrument
 * Type     : fm
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : FM drawbar organ
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FM_ORGAN_H
#define MB_FM_ORGAN_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFmOrgan : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-fm-organ";
    static constexpr const char* PLUGIN_NAME    = "MB FM Organ";
    static constexpr const char* PLUGIN_TYPE    = "fm";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float modIndex = 2f;  // range [0, 10]
    float percussive = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbFmOrgan() = default;
    ~MbFmOrgan() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.modIndex = std::clamp(params.modIndex, 0f, 10f);
        params.percussive = std::clamp(params.percussive, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB FM Organ
        return input;
    }
};

#endif // MB_FM_ORGAN_H

File: built-in plugins dsp/mb-fm-organ.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFmOrganPlugin: PluginDefinition = { id: 'mb-fm-organ', slug: 'mb-fm-organ', name: 'MB FM Organ', category: 'instrument', type: 'fm', version: '1.0.0', description: 'FM drawbar organ', author: 'Max Booster', grade: 'A', oscillators: [{ type: 'sine', detune: 0, gain: 0.7 }], envelope: { attack: 0.01, decay: 0.1, sustain: 0.9, release: 0.1 }, parameters: [{ id: 'modIndex', name: 'Mod Index', type: 'float', defaultValue: 2, minValue: 0, maxValue: 10, automatable: true }, { id: 'percussive', name: 'Percussive', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { modIndex: 2, percussive: 0.5, volume: 0.8 } };

export default MbFmOrganPlugin;
